python/addons/preprocessors.py

```diff
@@ -0,0 +1,176 @@
+
+        # Make a vocab list
+        vocab_list = [''] * (len(vocab) + 1)
+
+        for v, i in vocab.items():
+            vocab_list[i] = v
+
+        tok2index = tf.contrib.lookup.index_table_from_tensor(
+            tf.constant(vocab_list),
+            default_value=Offsets.UNK,
+            dtype=tf.string,
+            name='%s2index' % feature_name
+        )
+        return tok2index, vocab
+
+    def create_vocabs(self, model_base_dir, pid, features):
+        """
+        :model_file the path-like object to the model and model name.
+        :vocab_suffixes the list of vocab types. e.g. 'word', 'char', 'ner'.
+        """
+        indices = {}
+        vocabs = {}
+        for feature in features:
+            feature_vocab_file = os.path.join(model_base_dir, "vocabs-{}-{}.json".format(feature, pid))
+            if os.path.exists(feature_vocab_file):
+                indices[feature], vocabs[feature] = self._read_vocab(feature_vocab_file, feature)
+        return indices, vocabs
+
+    @staticmethod
+    def create_vectorizers(model_base_dir, pid):
+        """
+        :model_file the path-like object to the model and model name.
+        :vocab_suffixes the list of vocab types. e.g. 'word', 'char', 'ner'.
+        """
+        import pickle
+        return pickle.load(open(os.path.join(model_base_dir, "vectorizers-{}.pkl".format(pid)), "rb"))
+
+    def _assign_char_lookup(self):
+        upchars = tf.constant([chr(i) for i in range(65, 91)])
+        self.lchars = tf.constant([chr(i) for i in range(97, 123)])
+        self.upchars_lut = tf.contrib.lookup.index_table_from_tensor(mapping=upchars, num_oov_buckets=1, default_value=-1)
+
+    def _create_example(self):
+        serialized_tf_example = tf.placeholder(tf.string, name='tf_example')
+        feature_configs = {
+            self.FIELD_NAME: tf.FixedLenFeature(shape=[], dtype=tf.string),
+        }
+        tf_example = tf.parse_example(serialized_tf_example, feature_configs)
+        return tf_example
+
+    def run(self):
+        self._assign_char_lookup()
+        tf_example = self._create_example()
+        preprocessed = self.create_preprocessed_input(tf_example)
+        return tf_example, preprocessed
+
+    def create_preprocessed_input(self, tf_example):
+        """
+        Create a preprocessor chain inside of the tensorflow graph.
+        """
+        pass
+
+    @staticmethod
+    def reshape_indices(indices, shape):
+        reshaped = tf.sparse_reset_shape(indices, new_shape=shape)
+        # Now convert to a dense representation
+        x = tf.sparse_tensor_to_dense(reshaped)
+        x = tf.contrib.framework.with_shape(shape, x)
+        return x
+
+
+class Token1DPreprocessorCreator(PreprocessorCreator):
+
+    def __init__(self, model_base_dir, pid, features, **kwargs):
+        super(Token1DPreprocessorCreator, self).__init__(model_base_dir, pid, features, **kwargs)
+        self.mxlen = self.vectorizers['word'].mxlen
```

Comment:
This is still hard coded to `word` rather than being based on the `primary_key` idea from mead

Reply:
> I trained a sst2 model with this feature set, it was able to be exported but when I tried to use it with `classify-text.py` I got the error below.
> 
> ```json
>   "features": [
>     {
>       "name": "word",
>       "vectorizer": {
>         "type": "token1d",
> 	"transform": "baseline.lowercase"
>       },
>       "embeddings": {
>         "label": "w2v-gn"
>       }
>     },
>     {
>       "name": "char",
>       "vectorizer": {
>         "type": "char2d"
>       },
>       "embeddings": { "dsz": 30, "wsz": 30, "type": "char-conv" }
>     }
>   ],
> ```
> 
> ```
> python classify-text.py --model ../models/client/models/1/ --remote localhost:8500 --name test --text "I hate this movie so much" --preproc true
> ../models/client/models/1/model.assets
> ../models/client/models/1/vocabs-char-5506.json
> ../models/client/models/1/vocabs-word-5506.json
> Traceback (most recent call last):
>   File "classify-text.py", line 27, in <module>
>     for text, output in zip(texts, m.predict(texts, preproc=args.preproc)):
>   File "/home/blester/dev/work/baseline/python/baseline/services.py", line 190, in predict
>     outcomes_list = self.model.predict(examples)
>   File "/home/blester/dev/work/baseline/python/baseline/remote.py", line 161, in predict
>     outcomes_list = stub.Predict(request)
>   File "/home/blester/anaconda3/envs/dl/lib/python3.6/site-packages/grpc/_channel.py", line 500, in __call__
>     return _end_unary_response_blocking(state, call, False, None)
>   File "/home/blester/anaconda3/envs/dl/lib/python3.6/site-packages/grpc/_channel.py", line 434, in _end_unary_response_blocking
>     raise _Rendezvous(state, None, None, deadline)
> grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with (StatusCode.INVALID_ARGUMENT, You must feed a value for placeholder tensor 'char' with dtype int32 and shape [?,?,?]
> 	 [[{{node char}} = Placeholder[_output_shapes=[[?,?,?]], dtype=DT_INT32, shape=[?,?,?], _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]])>
> ```
> It seem that the way it is currently is coupled with the idea that the classifiers have one feature [(word)](https://github.com/dpressel/baseline/blob/50c7a692bc6fb668e78820ef1e244ffcfc1cd556/python/addons/preprocessors.py#L130) and the taggers have two [(word and char)](https://github.com/dpressel/baseline/blob/50c7a692bc6fb668e78820ef1e244ffcfc1cd556/python/addons/preprocessors.py#L163)
> 
> I would think that these would be dynamic based on the features and the vecotrizers to allow for things like word and character classifiers
> 
> I get a similar error when I have word and glove as features, The placeholder for for glove in not filled and GRPC complains about it.

i think the i was going with the idea that if you have a custom model you write a preproc exporter for that. but this can be generalized a bit i think.

Reply:
IOW preproc extensions are there to handle the transform fns. so if in your prev eg you had a transform fn for char then it won't be possible to generalize for that.

Reply:
the code is updated to do this.